sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -88,5 +106,103 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+            private readonly TimeSpan _tokenRefreshOffset;
+
+            private TokenState _tokenState;
+            private string? _headerValue;
+            private TaskCompletionSource<string>? _pendingTcs;
+            private DateTimeOffset _refreshOn;
+            private DateTimeOffset _expiresOn;
+
+            public AccessTokenCache(TimeSpan tokenRefreshOffset)
+            {
+                _tokenRefreshOffset = tokenRefreshOffset;
+            }
+
+            public string? SaveToken(AccessToken token, Exception? exception)
+            {
+                string? headerValue;
+                TaskCompletionSource<string>? pendingTcs;
+
+                lock (_syncObj)
+                {
+                    if (!string.IsNullOrEmpty(token.Token))
+                    {
+                        _headerValue = "Bearer " + token.Token;
+                        _refreshOn = token.ExpiresOn - _tokenRefreshOffset;
+                        _expiresOn = token.ExpiresOn;
+                        _tokenState = TokenState.Valid;
+                    }
+                    else
+                    {
+                        _headerValue = default;
+                        _refreshOn = default;
+                        _expiresOn = default;
+                        _tokenState = TokenState.Invalid;
+                    }
+
+                    pendingTcs = _pendingTcs;
+                    headerValue = _headerValue;
+                    _pendingTcs = null;
+                }
+
+                if (pendingTcs != null)
+                {
+                    if (headerValue != null)
+                    {
+                        pendingTcs.SetResult(headerValue);
+                    }
+                    else
+                    {
+                        pendingTcs.SetException(exception ?? new InvalidOperationException());
+                    }
+                }
+
+                return headerValue;
+            }
+
+            public (string? headerValue, Task<string>? pendingTask) UpdateTokenState()
+            {
+                lock (_syncObj)
+                {
+                    if (DateTimeOffset.UtcNow >= _expiresOn && _tokenState != TokenState.Pending)
+                    {
+                        _tokenState = TokenState.Pending;
+                        _headerValue = null;
+                        return (null, null);
+                    }
+
+                    if (DateTimeOffset.UtcNow >= _refreshOn && _tokenState == TokenState.Valid)
+                    {
+                        _tokenState = TokenState.AboutToExpire;
+                        return (null, null);
+                    }
+
+                    switch (_tokenState) {
+                        case TokenState.Pending:
+                            _pendingTcs ??= new TaskCompletionSource<string>(TaskCreationOptions.RunContinuationsAsynchronously);
+                            return (null, _pendingTcs.Task);
```

Comment:
Who completes this TCS?

Reply:
`UpdateTokenState()`: https://github.com/Azure/azure-sdk-for-net/pull/13087/files#diff-8c190456ad69469f8f9e1ba176df1b68R153